set_instance_tags/app.py

```diff
@@ -0,0 +1,127 @@
+import json
+import logging
+import re
+import traceback
+
+import boto3
+import requests
+
+from crhelper import CfnResource
+
+
+MISSING_INSTANCE_ID_ERROR_MESSAGE = 'InstanceId parameter is required'
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+helper = CfnResource(
+  json_logging=False, log_level='DEBUG', boto_level='INFO')
+
+
+def get_ec2_client():
+  return boto3.client('ec2')
+
+
+def get_instance_id(event):
+  '''Get the instance id from event params sent to lambda'''
+  resource_properties = event.get('ResourceProperties')
+  instance_id = resource_properties.get('InstanceId')
+  if not instance_id:
+    raise ValueError(MISSING_INSTANCE_ID_ERROR_MESSAGE)
+  return instance_id
+
+
+def get_instance_tags(instance_id):
+  '''Look up the instance tags'''
+  client = get_ec2_client()
+  response = client.describe_tags(
+    Filters = [{'Name': 'resource-id', 'Values': [instance_id]}]
+  )
+  log.debug(f'EC2 describe tags response: {response}')
+  tags = response.get('Tags')
+  if not tags or len(tags) == 0:
+    raise Exception(f'No tags returned, received: {response}')
+
+  # format tags to easily pass to create_tags
+  for tag in tags:
+    tag.pop("ResourceId")
+    tag.pop("ResourceType")
+
+  return tags
+
+
+def get_principal_id(tags):
+  '''Find the value of the principal arn among the EC2 tags'''
+  principal_arn_tag = 'aws:servicecatalog:provisioningPrincipalArn'
+  for tag in tags:
+    if tag.get('Key') == principal_arn_tag:
+      principal_arn_value = tag.get('Value')
+      principal_id = principal_arn_value.split('/')[-1]
+      return principal_id
+  else:
+    raise ValueError('Could not derive a provisioningPrincipalArn from tags')
+
+
+def get_synapse_email(synapse_id):
+  '''Use the synapse id to look up the synapse user name'''
+  synapse_url = f'https://repo-prod.prod.sagebase.org/repo/v1/userProfile/{synapse_id}'
```

Comment:
you should use the Synapse python client here.  See https://python-docs.synapse.org/build/html/Client.html#synapseclient.Synapse.getUserProfile

Reply:
is this a requirement?  i'm borrowing code,  `set_bucket_tags` function that @tthyer wrote, that already works.  if you feel strongly about this then I would like to implement it with a follow on PR to this change.

Reply:
This code (lines 67-70) duplicates code in the Synapse Python client so the best practice would be do just use the single command from that package.  We have a working example of using the client here: https://github.com/Sage-Bionetworks-IT/lambda-budgets.  I do not have a strong opinion whether the change is added to this PR or added in a follow-up PR.